File: September 2024/K Sized Subarray Maximum(27 Sept).js
// JAVA CODE

class Solution {
    // Function to find maximum of each subarray of size k.
    public ArrayList<Integer> max_of_subarrays(int k, int arr[]) {
        // Your code here
        ArrayList<Integer> ans = new ArrayList<>();
        Deque<Integer> dq = new ArrayDeque<>();
        for(int i=0;i<arr.length;i++){
            if(dq.size()!=0 && dq.getFirst()==i-k)dq.removeFirst();
            while(dq.size()!=0 && arr[dq.getLast()]<=arr[i])dq.removeLast();
            dq.add(i);
            if(i>=k-1)ans.add(arr[dq.getFirst()]);
        }
        return ans;
    }

}

File: September 2024/K_sized_subarray_Maximum(27 sept).cpp
vector<int> max_of_subarrays(int k, vector<int> &arr) {
        
        deque<int> dq;
        vector<int> ans;
        
        int n = arr.size();
        
        for(int i = 0; i < k; i++){
            
            while(!dq.empty() && arr[i] >= arr[dq.back()]){
                
                dq.pop_back();
            }
            dq.push_back(i);
        }
        // ans.push_back(dq.front());
        
        for(int i = k; i < n; i++){
            
            ans.push_back(arr[dq.front()]);
            
            while(!dq.empty() && dq.front() <= i-k){
                
                dq.pop_front();
            }
            
            while(!dq.empty() && arr[i] >= arr[dq.back()]){
                
                dq.pop_back();
            }
            dq.push_back(i);
            
        }
        if(!dq.empty()){
            
            ans.push_back(arr[dq.front()]);
        }
        
        
        return ans;
        // your code here
    }
};